python/cudf/cudf/core/groupby/groupby.py

```diff
@@ -1252,7 +1249,7 @@ def apply(self, function, *args, engine="cudf"):
           on the grouped chunk.
         args : tuple
             Optional positional arguments to pass to the function.
-        engine: {'cudf', 'jit'}, default 'cudf'
+        engine: {'cudf', 'jit'}, default 'auto'
```

Comment:
For my knowledge: Is there a performance cost to the fallback? i.e. Does the JIT attempt have measurable overhead?

Reply:
Yes, it does have measurable overhead. One way of measuring it is with
```python
import cudf
df = cudf.DataFrame({
    'a':[0,1,1],
    'b':[1,2,3]
})

def func(grp):
    # binops can't be jitted without refcounting
    return grp + grp

grouped = df.groupby('a')

import cProfile
cProfile.run('grouped.apply(func)', sort='cumtime')
```

For this I get 
```
        1    0.000    0.000    0.213    0.213 groupby_utils.py:207(_can_be_jitted)
```

Meaning it's quite impactful. However if this becomes a problem users should be able to obtain the old behavior by just passing `engine='cudf'`.

Reply:
Full disclosure, we anticipated this and [I was OK with it](https://github.com/rapidsai/cudf/issues/13103#issuecomment-1502096723). I think the tradeoff is generally worthwhile. If we think it isn't then I think we'd just stop this work and remove 'auto' altogether.

Also I just noticed that 'auto' is not listed in the set of valid `engine` arguments here.

Reply:
Just to clarify — your snippet above is saying the JIT attempt costs ~200ms? That sounds right to me. I would think a cache could also be used here if needed to prevent multiple failed attempts from paying the overhead for the same function.

I am supportive of this change, because when it does pay off, it’s a big win. Just want to make sure we’re putting in the appropriate amount of engineering effort to mitigate the downside risk.

Reply:
> Also I just noticed that 'auto' is not listed in the set of valid engine arguments here.

Fixed this.